compiler/rustc_borrowck/src/diagnostics/explain_borrow.rs

```diff
@@ -248,11 +249,27 @@ impl<'tcx> BorrowExplanation<'tcx> {
 
                         if let LocalInfo::BlockTailTemp(info) = local_decl.local_info() {
                             if info.tail_result_is_ignored {
+                                // #133941: Don't suggest adding a semicolon if we can
+                                // confirm the expression already ends with one.
+                                let suggest_semicolon = {
+                                    let source_map = tcx.sess().source_map();
+
+                                    let hi = info.span.hi();
+                                    let span =
+                                        info.span.with_lo(hi).with_hi(hi + rustc_span::BytePos(1));
+
+                                    match source_map.span_to_snippet(span) {
+                                        Ok(span_str) => !span_str.contains(';'),
+                                        Err(_) => true,
+                                    }
```

Comment:
we probably want to change `BlockTailInfo::tail_result_is_ignored` instead. Using spans for such checks does not work in case the error involves macro calls. The current check may also have issues if something like `tail_expr ;` or `tail_expr/* */;` or `tail_expr<new-line>;` is used

I could imagine that we should either change `tail_result_is_ignored` to an enum tracking why it's ignored or alternatively change `LocalInfo::BlockTailTemp` to track whether the tail is an expr or a statement

Reply:
That makes sense! Thanks for the suggestion - I knew someone who knew more than me would steer me in a better direction. I'll look into that. 👍